File: app/src/main/java/com/petal/browser/view/AdapterSearch.java
package com.petal.browser.view;

import android.content.Context;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.preference.PreferenceManager;

import com.google.android.material.card.MaterialCardView;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import com.petal.browser.R;
import com.petal.browser.database.FaviconHelper;
import com.petal.browser.database.Record;
import com.petal.browser.unit.HelperUnit;


public class AdapterSearch extends BaseAdapter implements Filterable {
    private final Context context;
    private final int layoutResId;
    private final List<String> liveSuggestions = new ArrayList<>();
    private final List<CompleteItem> originalList;
    private List<CompleteItem> resultList;
    private int count = 0;
    private final CompleteFilter filter = new CompleteFilter();

    public synchronized void setLiveSuggestions(List<String> suggestions) {
        liveSuggestions.clear();
        if (suggestions != null) {
            liveSuggestions.addAll(suggestions);
        }
        filter.refilter();
    }

    public AdapterSearch(Context context, int layoutResId, List<Record> recordList) {
        this.context = context;
        this.layoutResId = layoutResId;
        this.originalList = new ArrayList<>();
        this.resultList = new ArrayList<>();
        getRecordList(recordList);
    }

    private void getRecordList(List<Record> recordList) {
        for (Record record : recordList) {
            if (record.getTitle() != null
                    && !record.getTitle().isEmpty()
                    && record.getURL() != null
                    && !record.getURL().isEmpty()) {
                originalList.add(new CompleteItem(record.getTitle(), record.getURL()));
            }
        }

        Set<CompleteItem> set = new HashSet<>(originalList);
        originalList.clear();
        originalList.addAll(set);
    }

    @Override
    public int getCount() {
        if (count > 0) {
            return resultList.size();
        } else {
            return 0;
        }
    }

    @Override
    public Filter getFilter() {
        return filter;
    }

    @Override
    public Object getItem(int position) {
        return resultList.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view = convertView;
        Holder holder;

        if (view == null) {
            view = LayoutInflater.from(context).inflate(layoutResId, null, false);
            holder = new Holder();
            holder.titleView = view.findViewById(R.id.titleView);
            holder.urlView = view.findViewById(R.id.dateView);
            holder.favicon = view.findViewById(R.id.item_icon);
            holder.albumCardView = view.findViewById(R.id.item_CardViewItem);
            view.setTag(holder);
        } else {
            holder = (Holder) view.getTag();
        }

        CompleteItem item = resultList.get(position);
        holder.titleView.setText(item.title);
        holder.urlView.setText(item.url);

        TypedValue typedValue = new TypedValue();
        context.getTheme().resolveAttribute(R.attr.colorSurfaceContainerHighest, typedValue, true);
        int color = typedValue.data;
        holder.albumCardView.setCardBackgroundColor(color);

        holder.favicon.setImageResource(R.drawable.icon_image_broken);
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(context);
        String s = sp.getString("searchInput", "");
        if (!s.isEmpty()) {
            HelperUnit.setHighLightedTextSearch(context, holder.urlView, s);
            HelperUnit.setHighLightedTextSearch(context, holder.titleView, s);
        }
        return view;
    }

    private static class CompleteItem {
        private final String title;
        private final String url;
        private int index = Integer.MAX_VALUE;

        private CompleteItem(String title, String url) {
            this.title = title;
            this.url = url;
        }

        String getTitle() {
            return title;
        }

        String getURL() {
            return url;
        }

        int getIndex() {
            return index;
        }

        void setIndex(int index) {
            this.index = index;
        }

        @Override
        public boolean equals(Object object) {
            if (!(object instanceof CompleteItem)) {
                return false;
            }
            CompleteItem item = (CompleteItem) object;
            return item.getTitle().equals(title) && item.getURL().equals(url);
        }

        @Override
        public int hashCode() {
            if (title == null || url == null) {
                return 0;
            }
            return title.hashCode() & url.hashCode();
        }
    }

    private static class Holder {
        private ImageView favicon;
        private TextView titleView;
        private TextView urlView;
        private MaterialCardView albumCardView;
    }

    private class CompleteFilter extends Filter {
        private CharSequence lastConstraint = "";

        public void refilter() {
            filter(lastConstraint);
        }

        @Override
        protected FilterResults performFiltering(CharSequence prefix) {
            if (prefix == null) {
                return new FilterResults();
            }
            lastConstraint = prefix;

            List<CompleteItem> workList = new ArrayList<>();
            Set<String> addedTitles = new HashSet<>();

            // 1. Add Live Google Search Recommendations
            synchronized (AdapterSearch.this) {
                for (String suggestion : liveSuggestions) {
                    if (suggestion != null && !suggestion.trim().isEmpty() && !addedTitles.contains(suggestion.toLowerCase())) {
                        CompleteItem sugItem = new CompleteItem(suggestion, suggestion);
                        sugItem.setIndex(-1); // Top priority
                        workList.add(sugItem);
                        addedTitles.add(suggestion.toLowerCase());
                    }
                }
            }

            // 2. Add matching local history & bookmarks
            for (CompleteItem item : originalList) {
                String titleLower = item.getTitle() != null ? item.getTitle().toLowerCase() : "";
                String urlLower = item.getURL() != null ? item.getURL().toLowerCase() : "";
                String prefixLower = prefix.toString().toLowerCase();

                if (titleLower.contains(prefixLower) || urlLower.contains(prefixLower)) {
                    if (!addedTitles.contains(titleLower)) {
                        int index = titleLower.indexOf(prefixLower);
                        if (index < 0) index = urlLower.indexOf(prefixLower);
                        item.setIndex(index >= 0 ? index + 10 : 100);
                        workList.add(item);
                        addedTitles.add(titleLower);
                    }
                }
            }

            workList.sort(Comparator.comparingInt(CompleteItem::getIndex));
            FilterResults results = new FilterResults();
            results.values = workList;
            results.count = workList.size();
            return results;
        }

        @SuppressWarnings("unchecked")
        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            count = results != null ? results.count : 0;
            if (results != null && results.count > 0) {
                resultList = (List<CompleteItem>) results.values;
                notifyDataSetChanged();
            } else {
                resultList = new ArrayList<>();
                notifyDataSetInvalidated();
            }
        }
    }
}

File: app/src/main/java/com/petal/browser/view/AdapterSearch.kt
package com.petal.browser.view

import android.content.Context
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.preference.PreferenceManager
import com.google.android.material.card.MaterialCardView
import com.petal.browser.R
import com.petal.browser.database.Record
import com.petal.browser.unit.HelperUnit
import java.util.ArrayList
import java.util.Comparator
import java.util.HashSet

class AdapterSearch(
    private val context: Context,
    private val layoutResId: Int,
    recordList: List<Record>
) : BaseAdapter(), Filterable {

    private val liveSuggestions: MutableList<String> = ArrayList()
    private val originalList: MutableList<CompleteItem> = ArrayList()
    private var resultList: List<CompleteItem> = ArrayList()
    private var count: Int = 0
    private val filter: CompleteFilter = CompleteFilter()

    init {
        getRecordList(recordList)
    }

    @Synchronized
    fun setLiveSuggestions(suggestions: List<String>?) {
        liveSuggestions.clear()
        if (suggestions != null) {
            liveSuggestions.addAll(suggestions)
        }
        filter.refilter()
    }

    private fun getRecordList(recordList: List<Record>) {
        for (record in recordList) {
            val title = record.title
            val url = record.getURL()
            if (!title.isNullOrEmpty() && !url.isNullOrEmpty()) {
                originalList.add(CompleteItem(title, url))
            }
        }

        val set: Set<CompleteItem> = HashSet(originalList)
        originalList.clear()
        originalList.addAll(set)
    }

    override fun getCount(): Int {
        return if (count > 0) {
            resultList.size
        } else {
            0
        }
    }

    override fun getFilter(): Filter {
        return filter
    }

    override fun getItem(position: Int): Any {
        return resultList[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val holder: Holder

        if (convertView == null) {
            view = LayoutInflater.from(context).inflate(layoutResId, null, false)
            holder = Holder()
            holder.titleView = view.findViewById(R.id.titleView)
            holder.urlView = view.findViewById(R.id.dateView)
            holder.favicon = view.findViewById(R.id.item_icon)
            holder.albumCardView = view.findViewById(R.id.item_CardViewItem)
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as Holder
        }

        val item = resultList[position]
        holder.titleView.text = item.title
        holder.urlView.text = item.url

        val typedValue = TypedValue()
        context.theme.resolveAttribute(R.attr.colorSurfaceContainerHighest, typedValue, true)
        val color = typedValue.data
        holder.albumCardView.setCardBackgroundColor(color)

        holder.favicon.setImageResource(R.drawable.icon_image_broken)
        val sp = PreferenceManager.getDefaultSharedPreferences(context)
        val s = sp.getString("searchInput", "") ?: ""
        if (s.isNotEmpty()) {
            HelperUnit.setHighLightedTextSearch(context, holder.urlView, s)
            HelperUnit.setHighLightedTextSearch(context, holder.titleView, s)
        }
        return view
    }

    private class CompleteItem(val title: String, val url: String) {
        var index: Int = Int.MAX_VALUE

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is CompleteItem) return false
            return title == other.title && url == other.url
        }

        override fun hashCode(): Int {
            return title.hashCode() xor url.hashCode()
        }
    }

    private class Holder {
        lateinit var favicon: ImageView
        lateinit var titleView: TextView
        lateinit var urlView: TextView
        lateinit var albumCardView: MaterialCardView
    }

    private inner class CompleteFilter : Filter() {
        private var lastConstraint: CharSequence = ""

        fun refilter() {
            filter(lastConstraint)
        }

        override fun performFiltering(prefix: CharSequence?): FilterResults {
            if (prefix == null) {
                return FilterResults()
            }
            lastConstraint = prefix

            val workList: MutableList<CompleteItem> = ArrayList()
            val addedTitles: MutableSet<String> = HashSet()

            // 1. Add Live Google Search Recommendations
            synchronized(this@AdapterSearch) {
                for (suggestion in liveSuggestions) {
                    if (suggestion.isNotBlank() && !addedTitles.contains(suggestion.lowercase())) {
                        val sugItem = CompleteItem(suggestion, suggestion)
                        sugItem.index = -1 // Top priority
                        workList.add(sugItem)
                        addedTitles.add(suggestion.lowercase())
                    }
                }
            }

            // 2. Add matching local history & bookmarks
            for (item in originalList) {
                val titleLower = item.title.lowercase()
                val urlLower = item.url.lowercase()
                val prefixLower = prefix.toString().lowercase()

                if (titleLower.contains(prefixLower) || urlLower.contains(prefixLower)) {
                    if (!addedTitles.contains(titleLower)) {
                        var index = titleLower.indexOf(prefixLower)
                        if (index < 0) index = urlLower.indexOf(prefixLower)
                        item.index = if (index >= 0) index + 10 else 100
                        workList.add(item)
                        addedTitles.add(titleLower)
                    }
                }
            }

            workList.sortBy { it.index }
            val results = FilterResults()
            results.values = workList
            results.count = workList.size
            return results
        }

        @Suppress("UNCHECKED_CAST")
        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            count = results?.count ?: 0
            if (results != null && results.count > 0) {
                resultList = results.values as List<CompleteItem>
                notifyDataSetChanged()
            } else {
                resultList = ArrayList()
                notifyDataSetInvalidated()
            }
        }
    }
}
